Add Defer instances for Decoder and Encoder

morgen-peschke opened the pull request:
Implements #2122

morgen-peschke committed:
Add Defer instances for Decoder and Encoder

zarthross commented on modules/tests/shared/src/test/scala/io/circe/DecoderSuite.scala:
```diff
@@ -763,4 +764,6 @@ class DecoderSuite extends CirceMunitSuite with LargeNumberDecoderTestsMunit {
     assertEquals(result.swap.toOption.map(_.size), Some(2))
   }
+
+  checkAll("Defer[Decoder]", DeferTests[Decoder].defer[MiniInt])
```
Can you add some property tests for recursive datastructures like what was used in the example?  Just to verify further the stack safety and serve as an example for users on how to user defer?

morgen-peschke replied:
I added non-property test, as generating recursive data structures are kind of annoying to write `Gen`s for.

Not impossible though, so if you feel strongly that they need to be property tests, I can take a pass at converting them.

zarthross replied:
I think we need this test to have some depth to it.  We need to have the test ensure the stack safety of the implementation of defer.  Its probably fine to use `List` as the type the way you did, but using a generated list of some significant length should do the trick.

morgen-peschke replied:
Clarification question: are you concerned about the stack safety of the `Defer` instance, or the `Decoder` it's used to write?

The former is tested by the `Defer` laws: 
```
  + Defer[Decoder]: defer.defer Identity 0.156s
  + Defer[Decoder]: defer.defer does not evaluate 0.001s
  + Defer[Decoder]: defer.defer is stack safe 0.319s
  + Defer[Decoder]: defer.defer matches fix 0.429s
```

The later I think may be out of our control.

zarthross commented on modules/core/shared/src/main/scala/io/circe/Decoder.scala:
```diff
@@ -1433,11 +1443,27 @@ object Decoder
     }
 
+  private final case class DeferredDecoder[A](decoder: () => Decoder[A]) extends Decoder[A] {
+    override def apply(c: HCursor): Result[A] = decodeAccumulating(c).toEither.leftMap(_.head)
```
I know it seems redundant, but can you reimplement the function you have in `decodeAccumulating` here as well?  The reason is `apply`/non-accumulating should fail FAST/on the first error, while this implementation still runs all the decoders before failing, but only returns the first error.

morgen-peschke replied:
This was a good opportunity to add the caching that was brought up in [typelevel/cats#4414](https://github.com/typelevel/cats/pull/4414)

zmccoy commented:
Could we add some documentation as well to the docs for this?  You could easily transform what you have in the accommodating issue and it would be great for discoverability of this.

morgen-peschke committed:
Defer[Decoder/Encoder] performance improvements.

- Cache the resolution loop so we don't have to do it every time the instance is called.
- Preserve short-circuit decoding behavior when not accumulating errors

morgen-peschke committed:
Add tests for Defer[Decoder/Encoder]

morgen-peschke committed:
Scalafmt catchup

zarthross commented:
@morgen-peschke Thanks for adding the tests.  It looks like this build failed MIMA for some reason, do you mind taking a look?

armanbilge commented on modules/core/shared/src/main/scala/io/circe/Decoder.scala:
```diff
@@ -1433,11 +1443,27 @@ object Decoder
    */
-  implicit final val decoderInstances: SemigroupK[Decoder] with MonadError[Decoder, DecodingFailure] =
-    new SemigroupK[Decoder] with MonadError[Decoder, DecodingFailure] {
+  implicit final val decoderInstances
+    : SemigroupK[Decoder] with MonadError[Decoder, DecodingFailure] with Defer[Decoder] =
+    new SemigroupK[Decoder] with MonadError[Decoder, DecodingFailure] with Defer[Decoder] {
```
I think you'll need to do something like this to fix bincompat

```suggestion
  @deprecated
  def decoderInstances: MonadError[Decoder, DecodingFailure] = decoderInstancesBinCompat
  
  implicit final val decoderInstancesBinCompat = 
    : SemigroupK[Decoder] with MonadError[Decoder, DecodingFailure] with Defer[Decoder] =
    new SemigroupK[Decoder] with MonadError[Decoder, DecodingFailure] with Defer[Decoder] {
```

zarthross replied:
Alternatively, we could just make a separate val for the `Defer` couldn't we?  `Defer` has no parents and no children so there shouldn't be any real need for it to be part of `decoderInstances` right?

armanbilge replied:
Sure, good point!

zarthross replied:
Also, does the order matter on those `with` clauses?  Maybe moving `Defer[Decoder]` to before `MonadError[Decoder, DecodingFailure]` instead of after?

armanbilge replied:
I'm not sure. But I do know that Scala 2 and Scala 3 have different algorithms for this, so it may be impossible to reconcile in a way to make all versions happy. We had a similar issue in Cats (actually we accidentally broke bincompat on Scala 3 since weren't checking it back then ... )

armanbilge replied:
Ref:
- https://github.com/typelevel/cats/issues/4062#issuecomment-984064544

morgen-peschke committed:
Move Defer[Decoder] to a separat instance

Hopefully this should fix the bincompat issue

morgen-peschke committed:
Merge branch 'series/0.14.x' into provide-defer-instances

morgen-peschke committed:
Add extra test for Decoder.recursive stack safety

morgen-peschke committed:
Add first pass of Defer.recursive docs

hamnis commented:
would you mind fixing the merge conflics here?

hamnis committed:
Merge branch 'series/0.14.x' into provide-defer-instances

hamnis committed:
formatting

hamnis commented:
failure is genuine: https://github.com/circe/circe/actions/runs/9078292566/job/24945019131?pr=2123#step:13:2785

johnynek commented on modules/tests/shared/src/test/scala/io/circe/DecoderSuite.scala:
```diff
@@ -779,4 +780,55 @@ class DecoderSuite extends CirceMunitSuite with LargeNumberDecoderTestsMunit {
+    implicit def uglyListDecoder[A: Decoder]: Decoder[List[A]] =
+      Decoder.recursive[List[A]] { implicit recurse =>
+        Decoder.instance { c =>
+          (c.downField("car").as[A], c.downField("cdr").as[Option[List[A]]]).mapN(_ :: _.getOrElse(Nil))
```
I don't think this is stack safe, and it looks like the tests fail.

So, recurse is doing to be used at `c.downField("cdr").as[Option[List[A]]]` but that's nested inside of a `mapN`. While nesteding `Defer` is stack safe in the construction you have here, composing with map2, which is what is happening here, isn't. Something like: `lazy val d = map2(a, defer(d))(_ :: _)`, `map2` would need to somehow unroll that loop. It's not easy to see how that could happen.

So, I think the stack depth will be the same depth as the object graph.

cats-parse has the same issue on parsing (which is the same basic typeclass as Decoder).

johnynek replied:
If you ask me, we should merely add a comment to recursive and remove this test. Many structures have depth = log(size), so they are perfectly safe to parse with this. If you have a linear depth object (such as a linked list), this isn't suitable if the lists are big.

morgen-peschke replied:
Something like this?

> Note: while a `Decoder` written using `Decoder.recursive` can prevent unneeded creation of `Decoder` instances when recursing, the resulting `Decoder` cannot be guaranteed to be stack-safe.

johnynek replied:
yeah, something like that seems right.

morgen-peschke committed:
Remove test incorrectly asserting result of Decoder.recursive will be stack stafe

johnynek commented on modules/core/shared/src/main/scala/io/circe/Encoder.scala:
```diff
@@ -23,40 +23,39 @@ import cats.data.Chain
+import java.time.ZoneOffset
+import java.time.ZonedDateTime
+import java.time.format.DateTimeFormatter
+import java.time.format.DateTimeFormatter.ISO_LOCAL_DATE_TIME
```
personally, I don't like these import reformats. I think grouped imports make it easier to see the package level dependencies and I would rather use `{ }` to the maximum vs the minimum.

morgen-peschke replied:
I don't have strong opinions on it, other than "let the tooling handle it" 😅 

In this case, it was the result of running `scalafixAll` and `scalafmtAll` (then reverting the changes to the files I hadn't touched).

morgen-peschke replied:
I can revert the changes manually if you find it particularly objectionable, but it might be worth revisiting the formatting config if that is the case 🤷🏻

satorg replied:
Also, in my experience, unfolded imports tend to trigger less conflicts on merges and rebases. I guess this could one of the reasons why it was made a default behavior for the OrganizeImports rule in scalafix.

morgen-peschke replied:
I needed to re-run the CI, so I've rolled back the import reformat

morgen-peschke committed:
Clean up mdoc errors

morgen-peschke committed:
Rollback changes to imports

morgen-peschke committed:
Merge branch 'series/0.14.x' into provide-defer-instances

morgen-peschke committed:
Expand the tutorial docs on recursive ADTs

morgen-peschke committed:
Correct instructions for generating microsite

morgen-peschke committed:
Merge branch 'provide-defer-instances' of https://github.com/morgen-peschke/circe into provide-defer-instances
